### PR TITLE
llvmPackages_14: Fix build on aarch64-linux

### PR DESCRIPTION
**Pull request by avdv**
###### Description of changes

This PR fixes building the LLVM packages on aarch64-linux.

The libcxx library needs a C++20 capable compiler toolchain since version 14.

Also, work around the linking problems described in https://github.com/NixOS/nixpkgs/issues/201254

ZHF: #199919

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by risicle** [comment]
@ofborg eval

File: pkgs/development/compilers/llvm/14/llvm/default.nix [hunk 1]
```diff
@@ -129,6 +129,9 @@ in stdenv.mkDerivation (rec {
   # E.g. mesa.drivers use the build-id as a cache key (see #93946):
   LDFLAGS = optionalString (enableSharedLibraries && !stdenv.isDarwin) "-Wl,--build-id=sha1";
 
+  # workaround https://github.com/NixOS/nixpkgs/issues/201254
+  NIX_LDFLAGS = optionalString (stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU) "-lgcc";
+
```
Comment (avdv) [inline pkgs/development/compilers/llvm/14/llvm/default.nix:134]: I suppose this change might be the reason that this would cause so many rebuilds, even on Darwin which should really not be effected at all.

Should I change that to add the `NIX_LDFLAGS` attribute optionally?

**Comment by risicle** [comment]
Hmm I'm still getting

```
/nix/store/1g5g7vzlcprmg1d6864fs5r4nf1h5b5j-binutils-2.39/bin/ld: CMakeFiles/clang_rt.scudo_standalone-dynamic-aarch64.dir/linux.cpp.o: in function `bool scudo::atomic_compare_exchange_strong<scudo::atomic_u32>(scudo::atomic_u32 volatile*, scudo::atomic_u32::Type*, scudo::atomic_u32::Type, scudo::memory_order)':
/build/compiler-rt-src-14.0.6/compiler-rt/lib/scudo/standalone/atomic_helpers.h:119: undefined reference to `__aarch64_cas4_acq'
```

for `compiler-rt-libc-14.0.6`

**Comment by avdv** [comment]
@risicle you're right, I have added the same workaround for compiler-rt and clang.

File: pkgs/development/compilers/llvm/14/llvm/default.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/development/compilers/llvm/14/llvm/default.nix:134]: You can avoid rebuilds by using
```nix
{
  NIX_LDFLAGS = if stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU then "-lgcc" else null;
}
```

Comment (avdv) [reply pkgs/development/compilers/llvm/14/llvm/default.nix:134]: Thank you, @vcunat! Changed.

**Comment by avdv** [comment]
When building `llvmPackages_14.libcxxabi`, I ran into this error:
```
n file included from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/strong_order.h:12,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/weak_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/partial_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/compare_partial_order_fallback.h:13,                   
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/compare:144,                                         
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/utility:236,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__functional_base:26,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/memory:808,         
                 from /build/libcxxabi-src-14.0.6/libcxxabi/../libcxx/src/include/atomic_support.h:13,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/src/cxa_default_handlers.cpp:19:
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h: In function 'constexpr _ToType std::__1::bit_cast(const _FromType&)':
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h:31:38: error: expected primary-expression before ',' token
   31 |     return __builtin_bit_cast(_ToType, __from);
      |                                      ^                                                                                                                                   
In file included from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/strong_order.h:12,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/weak_order.h:14,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/partial_order.h:14,                                             
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__compare/compare_partial_order_fallback.h:13,                
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/compare:144,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/utility:236,
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__functional_base:26,        
                 from /nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/memory:808,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/../libcxx/src/include/atomic_support.h:13,
                 from /build/libcxxabi-src-14.0.6/libcxxabi/src/cxa_default_handlers.cpp:19:         
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h: In function 'constexpr _ToType std::__1::bit_cast(const _FromType&)':
/nix/store/87sa00y1dyyfyigqzav5cjjgzrr44rx4-libcxx-headers-14.0.6/include/c++/v1/__bit/bit_cast.h:31:38: error: expected primary-expression before ',' token
   31 |     return __builtin_bit_cast(_ToType, __from);                                                                                                                          
      |                                      ^                                                                                                                                   
make[2]: *** [src/CMakeFiles/cxxabi_static.dir/build.make:90: src/CMakeFiles/cxxabi_static.dir/cxa_default_handlers.cpp.o] Error 1               
make[1]: *** [CMakeFiles/Makefile2:165: src/CMakeFiles/cxxabi_static.dir/all] Error 2                                                                                            
make[1]: *** Waiting for unfinished jobs....                                                                                                                                     
make[2]: *** [src/CMakeFiles/cxxabi_shared.dir/build.make:90: src/CMakeFiles/cxxabi_shared.dir/cxa_default_handlers.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....                                                                    
```
According to https://github.com/llvm/llvm-project/issues/56828 this happens with GCC 10, but works OK with GCC 11.

**Comment by risicle** [comment]
Linking error again for `lldb`

File: pkgs/development/compilers/llvm/14/clang/default.nix [hunk 2]
```diff
@@ -26,6 +26,9 @@ let
 
     buildInputs = [ libxml2 libllvm ];
 
+    # workaround https://github.com/NixOS/nixpkgs/issues/201254
+    NIX_LDFLAGS = if stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU then "-lgcc" else null;
```
Comment (ehmry) [inline pkgs/development/compilers/llvm/14/clang/default.nix:30]: We should consider if adding an argument above like `{ withLibgcc ? stdenv.isLinux && stdenv.isAarch64 && stdenv.cc.isGNU, ... }:` is better than putting these conditions inline in the body of derivations. This could make overriding LLVM easier and I think its self-documentation to have code in the form of `effect ? cause` rather than `if ${cause} then ${action} else null`.

**Comment by avdv** [comment]
> Linking error again for `lldb`

Fixed.

I tried to use nixpkgs-review, but this never finishes... :shrug:

File: pkgs/development/compilers/llvm/14/clang/default.nix (hunk 2 above)
Comment (avdv) [reply pkgs/development/compilers/llvm/14/clang/default.nix:38]: Since this is just an ugly workaround, I would prefer to keep this as ugly as possible. :grimacing:

File: pkgs/development/compilers/llvm/14/clang/default.nix [hunk 3]
```diff
@@ -1,6 +1,14 @@
...
+, stdenv
+, llvm_meta
+, monorepoSrc
+, runCommand
+, substituteAll
+, cmake
+, libxml2
+, libllvm
+, version
+, python3
```
Comment (primeos) [inline pkgs/development/compilers/llvm/14/clang/default.nix:11]: I'm not a fan of reformatting code in commits like `llvmPackages_14: Add -lgcc to NIX_LDFLAGS for lldb`. IMO that belongs in a separate commit (especially since https://github.com/NixOS/nixpkgs/pull/201485/commits/b93b4f60855d15fc7784b5fc5c57cb0c3cd899b2 doesn't seem to make any functional changes in this file at all so maybe this change was added by accident?).

**Review by primeos — CHANGES_REQUESTED** [review]
Just as a TODO: The commit history needs to be cleaned up (one commit per logical change - currently there are basically a few fixup commits - I'd suggest to either put everything into one commit or one to fix the build and one for the `-lgcc` changes (the latter would probably be better)).

Ideally this should also target `llvmPackages_git` as well (normally that would be required but `llvmPackages_git` is in bad shape since many PRs forget to target it and `llvmPackages_15` already has a PR that is stuck in limbo...).

File: pkgs/development/compilers/llvm/14/clang/default.nix (hunk 3 above)
Comment (avdv) [reply pkgs/development/compilers/llvm/14/clang/default.nix:11]: Yes, this was added by accident. Auto-formatted file + `git commit -am ...`.

Comment (avdv) [reply pkgs/development/compilers/llvm/14/clang/default.nix]: fixed.

**Comment by avdv** [comment]
@primeos  I have rebased onto master and refined the commit history.

**Review by primeos — APPROVED** [review]
LGTM now, thanks! :)

**Comment by vcunat** [comment]
This can't go directly to master.  OfBorg doesn't check this for aarch64-linux, but it is a large rebuild:
```
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
  14523 aarch64-linux
```

**Comment by mweinelt** [comment]
Bisected a broken `llvm_14` on aarch64-linux to 74965741696bf79fd67501a55b1b60683c2b5d62.

In https://hydra.nixos.org/build/201827173 it fails >80 tests.

**Comment by ehmry** [comment]
Go ahead and revert my merge if it doesn't make things worse.

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -14749,6 +14749,8 @@ with pkgs;
     buildLlvmTools = buildPackages.llvmPackages_14.tools;
     targetLlvmLibraries = targetPackages.llvmPackages_14.libraries or llvmPackages_14.libraries;
     targetLlvm = targetPackages.llvmPackages_14.llvm or llvmPackages_14.llvm;
+  } // lib.optionalAttrs (stdenv.hostPlatform.isAarch64 && stdenv.hostPlatform.isLinux && buildPackages.stdenv.cc.isGNU) {
+    stdenv = gcc11Stdenv;
```
Comment (Artturin) [inline pkgs/top-level/all-packages.nix:14753]: https://github.com/NixOS/nixpkgs/issues/108305

**Comment by avdv** [comment]
> Bisected a broken `llvm_14` on aarch64-linux to [7496574](https://github.com/NixOS/nixpkgs/commit/74965741696bf79fd67501a55b1b60683c2b5d62).
> 
> In [hydra.nixos.org/build/201827173](https://hydra.nixos.org/build/201827173) it fails >80 tests.

It fails all the gold tests. Which is caused by:
```
/nix/store/cwivsnh1p591x7yydynrz0xc1zcq54p3-binutils-2.39/bin/ld.gold: error: /build/llvm-src-14.0.6/llvm/build/lib/LLVMgold.so: could not load plugin library: /nix/store/kyrb20x8ldfhih953ffd8yba42qfb5xy-gcc-9.5.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /build/llvm-src-14.0.6/llvm/build/lib/libLLVM-14.so)
```
Would it be feasible to disable the gold linker plugin on this platform?
